src/hotspot/share/cds/cppVtables.cpp

```diff
@@ -322,5 +318,6 @@ void CppVtables::zero_archived_vtables() {
 
 bool CppVtables::is_valid_shared_method(const Method* m) {
   assert(MetaspaceShared::is_in_shared_metaspace(m), "must be");
-  return vtable_of(m) == _index[Method_Kind]->cloned_vtable();
+  return vtable_of(m) == _index[Method_Kind]->cloned_vtable() ||
+         vtable_of(m) == _archived_cpp_vtptrs[Method_Kind];
```

Comment:
I am not sure if this needs any fixing. If `m` is in the archive (as the above assert says), then its vtable should always be `_index[Method_Kind]->cloned_vtable()`.

Reply:
`_index` is updated to use the runtime vtables after `CppVtables::dumptime_init()`, so we must check `_archived_cpp_vtptrs` as well. This is a new condition that can happen when dumping the final archive.

Reply:
> _index is updated to use the runtime vtables after CppVtables::dumptime_init(), so we must check _archived_cpp_vtptrs as well.

That's right. I agree with your point, but I think the method `is_valid_shared_method` is not used during dump time. One of the caller is `Method::restore_unshareable_info` which is invoked during archive loading, and the other caller is related to JNI APIs to call Java methods. So the context of the call makes me think the current condition is sufficient. But again, it can be argued that this function can break if it gets used in some other context where the current condition is not sufficient.
I think I am fine with this change.